bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt

```diff
@@ -0,0 +1,124 @@
+                .addOutputState(SampleState(ourIdentity), SampleContract.CONTRACT_NAME)
+                .addCommand(SampleContract.Commands.Issue(), ourIdentity.owningKey)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
+
+        builder.verify(serviceHub)
+        val stx = serviceHub.signInitialTransaction(builder)
+        return subFlow(FinalityFlow(stx, listOf()))
+    }
+}
+
+@InitiatingFlow
+class TransferAssetFlow(val sampleState : StateAndRef<SampleState>,
+                              val bno : Party,
+                              val partyToTransferTo : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        if (ourIdentity != sampleState.state.data.owner)
+            throw FlowException("Only owner of the state can transfer it")
+        if (ourIdentity == partyToTransferTo)
+            throw FlowException("sender and recipient should be different parties")
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // get the membership from the node's membership cache
+        val membershipsCacheHolderService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
+        val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity)
+
+        if (ourMembership == null) {
+           throw FlowException("Invalid membership: ourMembership is null")
+        }
+
+        val builder = TransactionBuilder(notary)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
+                .addInputState(sampleState)
+                .addOutputState(SampleState(partyToTransferTo))
+                .addCommand(SampleContract.Commands.Transfer(), ourIdentity.owningKey, partyToTransferTo.owningKey)
+
+        // create a session
+        val counterPartySession = initiateFlow(partyToTransferTo)
+        // requesting other party to send us their billing states
+        counterPartySession.send(RequestMembershipForTransaction(bno))
+
+
+        val counterPartyMembershipState = subFlow(ReceiveStateAndRefFlow<MembershipState<Any>>(counterPartySession)).single()
+        builder.addReferenceState(ReferencedStateAndRef(counterPartyMembershipState))
+
+        // verifying the transaction
+        builder.verify(serviceHub)
+
+        val selfSignedTx = serviceHub.signInitialTransaction(builder)
+
+        val allSignedTx = subFlow(CollectSignaturesFlow(selfSignedTx, listOf(counterPartySession)))
+        return subFlow(FinalityFlow(allSignedTx, listOf(counterPartySession)))
+    }
+}
+
+@InitiatedBy(TransferAssetFlow::class)
+class TransferAssetFlowResponder(val session : FlowSession) : FlowLogic<Unit>() {
+    @Suspendable
+    override fun call() {
+
+        // receiving request for billing state
+        val bno = session.receive<RequestMembershipForTransaction>().unwrap { it }
```

Comment:
It's important to always verify things that you are received from the wire. For example, should we verify that the received identity is indeed a legit BNO?

Reply:
good point. done